ADR/0039-spdx-support.md

```diff
@@ -0,0 +1,146 @@
+
+| SPDX Attribute                  | CycloneDX Attribute                                        |
+|---------------------------------|------------------------------------------------------------|
+| Formulations.components = [{}]  | Relationships = [{                                         |
+|                                 |  spdxElementId = <A-PACKAGE-ID>                            |
+|                                 |  relationshipType=BUILD_TOOL_OF                            |
+|                                 |  relatedSpdxElement=<ROOT-DOCUMENT-ID>                     |
+|                                 | }]                                                         |
+
+
+#### Metadata.tools
+The CycloneDX metadata.tools sub attributes that we are mostly interested in are the vendor and name elements. Information about the creation of the SPDX document can be stored into creationInfo. CreationInfo.creators element is basically a list of strings. There’s a vague specification about how it should be structured in the standard. Strings should be formatted in the following way: “<Attribute>: <Value>”. For example vendor should be stored as “Vendor: <vendor>”
+
+| SPDX Attribute                                 | CycloneDX Attribute                               |
+|------------------------------------------------|---------------------------------------------------|
+| Metadata.tools = [{“vendor”: “X”, “name”: “Y”] | CreationInfo.creators = [“Vendor: X”, “Tool: Y”]  |
+
+
+#### Merging SPDX
+##### Packages
+Packages of two SPDX documents can be merged together as a concatenation of two lists. In cycloneDX component elements can have only a single purl attribute, therefore component elements representing packages with the same name and version but with different purl have to be stored as multiple elements. CycloneDX package elements can bear multiple purls. Therefore multiple cycloneDX components can be squashed together into single SPDX package element with purls concatenated into a single list. Following rules are applied to packages merging process:
+- Packages with the same package name and versionInfo are squashed into single package element
+- Package with the same package name and versionInfo set to None (or empty) is squashed with package with the same name and non-empty versionInfo
+NOTE: packages cannot be merged together based on SPDXID attribute as there’s no specification in the spdx standard on how SPDXID should be calculated. Individual tools can calculate it differently while still passing condition to make it unique across the whole document.
+##### Relationships
+SPDX relationships represent graph/tree structure of relations of elements in the document. The Root element is the SPDX document itself. Individual packages are in relationship CONTAINS with the root document. In the case of packages which were used to build the container, packages are in relationship BUILD_TOOL_OF with the root document.
```

Comment:
> Individual packages are in relationship CONTAINS with the root document.

Why is that important?

Reply:
I don't understand your question. It describes structure of the document so I guess that's why it's important.

Reply:
CycloneDX 1.4 probably can't, but we use 1.5

CycloneDX 1.5:
```jsonc
{
  "metadata": {
    "component": {
      // main component (e.g. the output container image)
    }
  },
  "components": [
    // the dependencies included in the container image
  ],
  "formulation": [
    {
      "components": [
        // build-time-only dependencies of the container image
      ]
    }
  ]
}
```

Conversion to SPDX, in my mind:

```
SPDXRef-DOCUMENT DESCRIBES .metadata.component
.metadata.component CONTAINS .components[]
.formulation[].components[] BUILD-TOOL-OF .metadata.component
```

Reply:
In the current state, Konflux SBOMs don't have a useful .metadata.component. Which IMO shouldn't stand in the way of this ADR, we should document the desired state and file a story to make it so

Reply:
so you suggest to create another virtual package to describe build-tool-of components? Btw, if you want to have that detailed level of relationships then I think right relationship would be:
SPDXRef-DOCUMENT  -> DESCRIBES:  SPDXRef-Image
SPDXRef-Image -> CONTAINS: PackageA
SPDXRef-DOCUMENT  -> DESCRIBES: SPDXRef-BaseImages
SPDXRef-BaseImages -> CONTAINS:  SPDXRef-BaseImage-Package-1
SPDXRef-BaseImages -> BUILD_TOOL_OF:  SPDXRef-Image

but that's quite complicated relations and in the end only relavant thing is

SPDXRef-BaseImage-Package-1 -> BUILD_TOOL_OF: SPDXRef-Image

And then there's no point of creating SPDXRef-Image as it would be basically equivalent of the virtual element which syft creates already

Reply:
> And then there's no point of creating SPDXRef-Image

There very much is! But it should not be a virtual element, it should be the output image. In fact I believe having a proper SPDXRef-Image is critical for another feature that @allda and team are working on https://github.com/konflux-ci/build-tasks-dockerfiles/pull/165#issuecomment-2416479615

Reply:
> SPDXRef-DOCUMENT -> DESCRIBES: SPDXRef-Image
> SPDXRef-Image -> CONTAINS: PackageA
> SPDXRef-BaseImages -> BUILD_TOOL_OF: SPDXRef-Image

These make sense to be. The other BaseImages relations not so much - the SBOM doesn't describe the base images, those have their own SBOMs

Reply:
@midnightercz I think virtual packages should be completely avoided unless they solve a real problem. (The kind of problem they _might_ solve: describing the build root for an RPM.)

I'm concerned about "base image" here: I think we're talking about builder images, not parent images.

And the view @chmeliik has here agrees with what Rogue and I discussed last week: we should include references (by purl) to the builder images, but we should not include references to their content.

Reply:
> I'm concerned about "base image" here: I think we're talking about builder images, not parent images.

Asi in, "parent image" = the last FROM instruction, "builder image" = any FROM instruction other than the last?

Currently, Konfux reports both of those in the `.formulations[]`, distinguished by a [custom property](https://github.com/konflux-ci/build-tasks-dockerfiles/blob/c34cbcd4274879871d0e968a78f9704b643ba7b3/sbom-utility-scripts/scripts/base-images-sbom-script/app/base_images_sbom_script.py#L52-L55)

Reply:
> Asi in, "parent image" = the last FROM instruction, "builder image" = any FROM instruction other than the last?

Exactly.

Reply:
> These make sense to me. The other BaseImages relations not so much - the SBOM doesn't describe the base images, those have their own SBOMs

So the you basically want:
```
SPDXRef-DOCUMENT -> DESCRIBES: SPDXRef-Image
SPDXRef-DOCUMENT -> CONTAINS: SPDXRef-Image
SPDXRef-Image -> CONTAINS: PackageA
SPDXRef-BuilderImage-Package-1-> BUILD_TOOL_OF: SPDXRef-Image
```
So for purl, I assume it will be something like:
`pkg:docker/ubi9/ubi@sha256:abcdef?repository_url=registry.redhat.io`

Reply:
That looks correct except for
> SPDXRef-DOCUMENT -> CONTAINS: SPDXRef-Image

Reply:
And the purl will be pk:oci rather than pkg:docker, same as the builder/base images (but the exact format doesn't matter that much for this ADR(. And FYI, Ales is already working on the purl https://github.com/konflux-ci/build-tasks-dockerfiles/pull/181